server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -286,8 +274,17 @@ private Updates changes(ShardId shardId) {
     /**
      * Remove allocation id of this shard from the set of in-sync shard copies
      */
-    private void removeAllocationId(ShardRouting shardRouting) {
-        changes(shardRouting.shardId()).removedAllocationIds.add(shardRouting.allocationId().getId());
+    void removeAllocationId(ShardRouting shardRouting) {
+        if (shardRouting.active()) {
+            if (shardRouting.primary()) {
+                Updates updates = changes(shardRouting.shardId());
+                if (updates.firstFailedPrimary == null) {
+                    // more than one primary can be failed (because of batching, primary can be failed, replica promoted and then failed...)
+                    updates.firstFailedPrimary = shardRouting;
```

Comment:
I think we should leave the "firstFailedPrimary" logic in the shardFailed method for now. I think that this logic can go away later.

Reply:
Yep, I put it back to the `shardFailed` method.